[DOC] Upgrades from older Strimzi versions

laidan6000 opened the pull request:
Signed-off-by: Daniel Laing <dlaing@redhat.com>

### Type of change
- Documentation

### Description

Guide(s) updated: _Deploying and Upgrading Strimzi_

This pull request makes further updates to the "Upgrading Strimzi" chapter based the redesigned Kafka upgrade process completed in PR #4314 .

Outstanding to do / discuss
- Add a new procedure for "Upgrading from an older Strimzi version". Older = older than `Strimzi latest - 1`.
- Is adding a new procedure the best approach or should we modify the existing procedures?

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [X] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [X] Reference relevant issue(s) and close them after merging
 - RELATED TO ISSUE **#1160** - Use Info level k8s `Event` to tell user about phases of Kafka upgrade
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

laidan6000 commented on documentation/assemblies/upgrading/assembly-upgrade.adoc:
```diff
@@ -6,10 +6,41 @@
+
+* The *Kafka versions* column lists the Kafka versions that are supported by each Strimzi version. 
+
+//When upgrading to an Operator version that drops support for the Kafka version you are upgrading from, the upgraded Cluster Operator returns an error message, for example:
```
TBC: Decide if this information is better here, or in the Kafka upgrade procedure.

laidan6000 commented on documentation/assemblies/upgrading/assembly-upgrade-kafka-versions.adoc:
```diff
@@ -6,18 +6,31 @@
+|===
+
+Additional rolling updates occur if the new version of Kafka requires a new ZooKeeper version.
+//check the above statement is still accurate
```
FYI - @scholzj

laidan6000 commented on documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc:
```diff
@@ -36,6 +36,16 @@ kubectl apply -f install/cluster-operator
++
+This error occurs when upgrading to an Operator version that drops support for the previous Kafka version. 
+You will upgrade to a supported Kafka version in the next procedure, when you upgrade all brokers to a later Kafka version.
 
```
At the moment, this is presented as "information only"? 
Is that right?
Or, does the user need to update `spec.Kafka.version` now?

scholzj commented on documentation/assemblies/upgrading/assembly-upgrade.adoc:
```diff
@@ -6,10 +6,42 @@
 = Upgrading Strimzi
 
-Strimzi can be upgraded from a lower version to a higher version to take advantage of new features and enhancements, performance improvements, and security options.
+Strimzi can be upgraded from a given version to a later version to take advantage of new features and enhancements, performance improvements, and security options.
```
I wonder if we should really use here the actual version number ... since as this stands, it applies for Strimzi 0.22.0. It does not apply to 0.21 if upgrading from 0.19. And we do not know how it applies to 0.23 in the future (I hope it will, but 0.23 will have its own docs with possible changes if there are any). Maybe that is why the `later` is so confusing for me.

laidan6000 replied:
Agreed. I've changed this to use the actual version number.

> Strimzi can be upgraded to version {ProductVersion} to take advantage of new features and enhancements, performance improvements, and security options.

scholzj commented on documentation/assemblies/upgrading/assembly-upgrade.adoc:
```diff
@@ -6,10 +6,42 @@
+For example, upgrading from Strimzi 0.20.1 to 0.21.0.
+
+Multi-version::
+Upgrading Strimzi from the current version to any later version within a single upgrade (skipping one or more intermediate versions).
```
Again, this is documentation for Strimzi 0.22. So this is about upgrading _from older versions to current version_ or possibly _from older versions to Strimzi {ProductVersion}_.

laidan6000 replied:
Changed to:

scholzj commented on documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc:
```diff
@@ -37,14 +37,29 @@ kubectl apply -f install/cluster-operator
++
+[source,shell,subs="+quotes,attributes"]
+----
+"Version {KafkaVersionLower} is not supported. Supported versions are: #...".
```
`{KafkaVersionLower}` is technically still supported so the message might be confusing. We should either add a new variable for the _unssuported_ version or just hardcode here something ... for example `2.4.0`.

laidan6000 replied:
I hardcoded the error message for Kafka 2.4.0. That's much simpler than using variables.

scholzj reviewed (commented):
I think technically it is good. It is just some of the wording which I commented about.

laidan6000 commented on documentation/modules/upgrading/proc-upgrade-brokers-newer-kafka.adoc:
```diff
@@ -6,7 +6,7 @@
-This procedure describes how to upgrade a Strimzi Kafka cluster to a higher version of Kafka.
+This procedure describes how to upgrade a Strimzi Kafka cluster to a later version of Kafka.
 
 .Prerequisites
```
I don't think this prerequisite is still true:

> The Cluster Operator, which supports both versions of Kafka, is up and running.

laidan6000 commented:
@scholzj - I addressed your comments, thank you. Also:

- Changed the ProductVersion attribute to `0.22.0`.

- Updated the terminology to make the content specific to Strimzi 0.22 as far as possible. As Jakub pointed out, there was a mix of version-specific conent and specific content.

- Now using "Strimzi version 0.22.0" (with the ProductVersion attribute). The phrase "new Strimzi version" is sometimes used instead.

- Now using "the latest supported Kafka version" instead of "a later version".

- Removed the `KafkaVersionEarlier` and `KafkaVersionLater` attributes because they add too much complexity.

scholzj commented on documentation/assemblies/upgrading/assembly-upgrade.adoc:
```diff
@@ -6,10 +6,40 @@
+Two upgrade paths are possible:
+
+Incremental:: 
+Upgrading Strimzi from version 0.21.1 to version {ProductVersion}.
```
We should write around the 0.21.1 release otherwise we would need to update it all the time. MAybe ...

```suggestion
Upgrading Strimzi from previous version to version {ProductVersion}.
```

laidan6000 replied:
OK, I guess it's likely to get missed. I'll change it to "the previous version".

tombentley replied:
But does previous version mean 0.20.0 or a 0.20.1? It's a bit ambiguous. Do we mean previous _minor_ version?

scholzj replied:
I guess that is good point @laidan6000 ^^^ ... _previous minor version_ would be better.

PaulRMellor replied:
Maybe an _(X.Y)_ example in

laidan6000 replied:
Changed to "the previous minor version". I tried an X.Y example but it seemed confusing.

tombentley commented on documentation/assemblies/upgrading/assembly-upgrade-kafka-cluster-operator.adoc:
```diff
@@ -7,7 +7,7 @@
 The steps to upgrade your Cluster Operator deployment to use Strimzi {ProductVersion} are described in this section.
 
-Follow this procedure if you deployed the Cluster Operator using the installation YAML files.
+Follow this procedure if you deployed the Cluster Operator using the installation YAML files rather than {OperatorHub}.
```
What about helm?

laidan6000 replied:
I think we'll need to open a new issue for Helm.

The Deploying and upgrading Strimzi guide should be updated to reflect deployment using Helm charts. The Upgrading Strimzi chapter needs to describe how to upgrade Strimzi if you deployed via Helm.

What do you think, @scholzj ?

scholzj replied:
I think that Kafka upgrades are the same regardless of OLM or Helm. For the operator upgrade, I'm not sure we have anyone who knows how to do it with Helm :-/

tombentley commented on documentation/modules/upgrading/proc-upgrade-brokers-newer-kafka.adoc:
```diff
@@ -6,19 +6,19 @@
-* The `Kafka.spec.kafka.config` does _not_ contain options that are not supported in the version of Kafka that you are upgrading to.
-* Whether the `log.message.format.version` and `inter.broker.protocol.version` for the current Kafka version need to be updated for the new version.
+* The `Kafka.spec.kafka.config` does _not_ contain options that are not supported in the new Kafka version.
+* Whether the `log.message.format.version` and `inter.broker.protocol.version` for the current Kafka version need to be updated for the new Kafka version.
```
I wonder whether we're phrasing this right. It's not like these things necessarily _need_ to be changed. Rather it's up to the user to manage them and thus understand what they do and so decide when it's right for them to change them.

laidan6000 replied:
I'm certainly not keen on the second bullet. How would the user find out which options are _not_ supported in the new Kafka version? By checking the Kafka release notes? Are we just covering our backs or is this an actual concern, I wonder.

As for the third bullet, I can rephrase the sentence and perhaps move it up to the introduction.

laidan6000 replied:
Line 17 is now part of the procedure introduction:

> Compared to your current Kafka version, the new version might support a higher _log message format version_ or _inter-broker protocol version_, or both. Follow the steps to upgrade these versions, if required. For more information, see Kafka versions.

tombentley commented on documentation/modules/upgrading/proc-upgrade-brokers-newer-kafka.adoc:
```diff
@@ -52,7 +52,9 @@ If the `log.message.format.version` and `inter.broker.protocol.version` are unse
 NOTE: The value of `log.message.format.version` and `inter.broker.protocol.version` must be strings to prevent them from being interpreted as floating point numbers.
 
-. Change the `Kafka.spec.kafka.version` to specify the new version (leaving the `log.message.format.version` and `inter.broker.protocol.version` at the _current_ versions).
+. Change the `Kafka.spec.kafka.version` to specify the new Kafka version.
```
Unless it's already explained elsewhere, I think it would be helpful to explain here that this change is about ensuring that all the brokers in the cluster are using the new binary, and during this change they need to continue using the old IBP so that the brokers can continue communicating.

laidan6000 replied:
Yes, I think that's useful to know at this stage. I added a note:

> NOTE: Changing the `kafka.version` ensures that all brokers in the cluster will be upgraded to start using the new broker binaries. During this process, some brokers are using the old binaries while others have already upgraded to the new ones. Leaving the `inter.broker.protocol.version` unchanged ensures that the brokers can continue to communicate with each other throughout the upgrade.

tombentley commented on documentation/modules/upgrading/proc-upgrade-brokers-newer-kafka.adoc:
```diff
@@ -83,39 +85,19 @@ See xref:con-versions-and-images-str[Kafka version and image mappings]
 
 . Save and exit the editor, then wait for rolling updates to complete.
 +
-NOTE: Additional rolling updates occur if the new version of Kafka has a new ZooKeeper version.
-+
-Check the update in the logs or by watching the pod state transitions:
-+
-[source,shell,subs=+quotes]
-----
-kubectl logs -f _CLUSTER-OPERATOR-POD-NAME_ | grep -E "Kafka version upgrade from KafkaVersion.*[0-9.]+.* to KafkaVersion.*[0-9.]+.*completed"
-----
```
Glad to see these go :-)

tombentley commented on documentation/modules/upgrading/proc-upgrade-brokers-newer-kafka.adoc:
```diff
@@ -83,39 +85,19 @@ See xref:con-versions-and-images-str[Kafka version and image mappings]
+* Configure the brokers to send messages using the inter-broker protocol version of the new version of Kafka
 +
-NOTE: Clients are still using the old version, so brokers will convert messages to the old version before sending them to the clients. To minimize this additional load, update the clients as quickly as possible.
+IMPORTANT: Clients are still using the old Kafka version, so brokers will convert messages to the old version before sending them to the clients. To minimize this additional load, update all client applications as soon as possible.
```
I think this is misleading. It's not the client version itself that matters. Message conversion can be caused by different message format between producer and topic (which, yes defaults to log.message.format.version, but can be overridden per topic), but also things like mismatched compression. Can we remove this? I think the following step makes it a bit clearer, since the strategies are explained in a bit more detail.

laidan6000 replied:
Removed the note.

PaulRMellor commented on documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc:
```diff
@@ -37,20 +37,35 @@ kubectl apply -f install/cluster-operator
+
+.. Edit the Kafka custom resource.
+
+.. Change the `spec.Kafka.version` property to a supported Kafka version.
```
```suggestion
.. Change the `spec.kafka.version` property to a supported Kafka version.
```

PaulRMellor reviewed (approved):
Great improvement. Clear instructions.

laidan6000 commented on documentation/assemblies/upgrading/assembly-upgrade-kafka.adoc:
```diff
@@ -23,7 +24,7 @@ Depending on your chosen upgrade strategy, after updating the channel, either:
 For more information on using {OperatorHub} to upgrade Operators, see the {OLMOperatorDocs}.
 
-. Upgrade all Kafka brokers and client applications to the latest Kafka version.
+. Upgrade all Kafka brokers and client applications to a later supported Kafka version.
```
Agreed. I've updated the wording to "latest supported Kafka version" in multiple places.

laidan6000 commented on documentation/assemblies/upgrading/assembly-upgrade-kafka.adoc:
```diff
@@ -5,9 +5,10 @@
+To upgrade brokers and clients without downtime, you _must_ complete the upgrade procedures in the following order:
 
-. Update your Cluster Operator to the latest Strimzi version.
+. Update your Cluster Operator to a later Strimzi version.
```
Done. "Update your Cluster Operator to a new Strimzi version."

laidan6000 commented:
@tombentley - Thanks for the review. The PR is now ready for further checks.
